File: app/src/main/java/com/robotnec/budget/core/persistence/dao/impl/CategoryDaoImpl.java
package com.robotnec.budget.core.persistence.dao.impl;

import com.robotnec.budget.app.util.Mapper;
import com.robotnec.budget.core.domain.Category;
import com.robotnec.budget.core.persistence.BudgetDatabase;
import com.robotnec.budget.core.persistence.dao.CategoryDao;
import com.robotnec.budget.core.persistence.schema.CategoryRecord;
import com.yahoo.squidb.data.SquidCursor;
import com.yahoo.squidb.sql.Property;

import java.util.List;

/**
 * @author zak <zak@swingpulse.com>
 */

public class CategoryDaoImpl extends SquidbDaoTemplate<Category, CategoryRecord> implements CategoryDao {

    public CategoryDaoImpl(BudgetDatabase database) {
        super(database);
    }

    @Override
    CategoryRecord fromCursor(SquidCursor<CategoryRecord> cursor) {
        CategoryRecord record = new CategoryRecord();
        record.readPropertiesFromCursor(cursor);
        return record;
    }

    @Override
    Class<CategoryRecord> getRecordClass() {
        return CategoryRecord.class;
    }

    @Override
    List<Category> map(List<CategoryRecord> tableModels) {
        return Mapper.INSTANCE.fromCategoryRecords(tableModels);
    }

    @Override
    CategoryRecord map(Category item) {
        return Mapper.INSTANCE.toRecord(item);
    }

    @Override
    Category map(CategoryRecord record) {
        return Mapper.INSTANCE.fromRecord(record);
    }

    @Override
    Property.LongProperty getIdProperty() {
        return CategoryRecord.ID;
    }
}

File: app/src/main/java/com/robotnec/budget/core/persistence/dao/impl/CategoryDaoImpl.kt
package com.robotnec.budget.core.persistence.dao.impl

import com.robotnec.budget.app.util.Mapper
import com.robotnec.budget.core.domain.Category
import com.robotnec.budget.core.persistence.BudgetDatabase
import com.robotnec.budget.core.persistence.dao.CategoryDao
import com.robotnec.budget.core.persistence.schema.CategoryRecord
import com.yahoo.squidb.data.SquidCursor
import com.yahoo.squidb.sql.Property

/**
 * @author zak zak@swingpulse.com>
 */

class CategoryDaoImpl(database: BudgetDatabase) : SquidbDaoTemplate<Category, CategoryRecord>(database), CategoryDao {

    override val recordClass: Class<CategoryRecord>
        get() = CategoryRecord::class.java

    override val idProperty: Property.LongProperty
        get() = CategoryRecord.ID

    override fun fromCursor(cursor: SquidCursor<CategoryRecord>): CategoryRecord {
        val record = CategoryRecord()
        record.readPropertiesFromCursor(cursor)
        return record
    }

    override fun map(tableModels: List<CategoryRecord>): List<Category> {
        return Mapper.fromCategoryRecords(tableModels)
    }

    override fun map(item: Category): CategoryRecord {
        return Mapper.toRecord(item)
    }

    override fun map(record: CategoryRecord): Category {
        return Mapper.fromRecord(record)
    }
}
